src/sinks/datadog/metrics/encoder.rs

```diff
@@ -502,6 +528,101 @@ fn encode_timestamp(timestamp: Option<DateTime<Utc>>) -> i64 {
     }
 }
 
+// Given the vector source type, return the OriginService value associated with that integration.
+//
+// In order to preserve consistent behavior, we intentionally don't set origin metadata
+// for the case where the Datadog Agent did not set it.
+// Some sources such as `kafka`, `nats`, `redis` for example, are only capable of receiving metrics
+// with the `native` or `native_json` codec. In such cases we intentionally do not set the origin
+// metadata here, because the true origin will have already been determined to be a pass-through.
+fn source_type_to_service(source_type: &str) -> Option<u32> {
+    if source_type == "apache_metrics" {
+        Some(17)
+    } else if source_type == "aws_ecs_metrics" {
+        Some(209)
+    } else if source_type == "eventstoredb_metrics" {
+        Some(210)
+    } else if source_type == "host_metrics" {
+        Some(211)
+    } else if source_type == "internal_metrics" {
+        Some(212)
+    } else if source_type == "mongodb_metrics" {
+        Some(111)
+    } else if source_type == "nginx_metrics" {
+        Some(117)
+    } else if source_type == "open_telemetry" {
+        Some(213)
+    } else if source_type == "postgresql_metrics" {
+        Some(128)
+    } else if source_type == "prometheus_remote_write" {
+        Some(214)
+    } else if source_type == "prometheus_scrape" {
+        Some(215)
+    } else if source_type == "statsd" {
+        Some(153)
+    } else {
+        None
+    }
+}
+
+/// Determine the correct Origin metadata values to use depending on if they have been
+/// set already upstream or not. The generalized struct `DatadogMetricOriginMetadata` is
+/// utilized in this function, which allows the series and sketch encoding to call and map
+/// the result appropriately for the given protocol they operate on.
+fn generate_origin_metadata(
+    maybe_pass_through: Option<&DatadogMetricOriginMetadata>,
+    maybe_source_type: Option<&Arc<String>>,
+) -> Option<DatadogMetricOriginMetadata> {
+    let origin_product_value = option_env!("ORIGIN_PRODUCT")
+        .map(|p| {
+            p.parse::<u32>()
+                .expect("Env var ORIGIN_PRODUCT must be an unsigned 32 bit integer.")
+        })
+        .unwrap_or(14);
```

Comment:
💬 This is a relatively expensive operation to execute for every metric event, as it requires a global lookup (locking) and string parsing. Can we compute this value once and pass it as an argument to the corresponding functions?